src/decoder.test.ts

```diff
@@ -44,13 +218,246 @@ describe('lib/decoder', () => {
+          // key: "name"
+          0x44, 0x6e, 0x61, 0x6d, 0x65, 0xe2,
+          // key: "en"
+          0x42, 0x65, 0x6e,
+          // value: "Foo"
+          0x43, 0x46, 0x6f, 0x6f,
+          // key: "zh"
+          0x42, 0x7a, 0x68,
+          // value: "人"
+          0x43, 0xe4, 0xba, 0xba,
+        ],
+        name: 'nested map',
+      },
+      {
+        expected: { languages: ['en', 'zh'] },
+        input: [
+          0xe1,
+          // key: "languages"
+          0x49, 0x6c, 0x61, 0x6e, 0x67, 0x75, 0x61, 0x67, 0x65, 0x73,
+          // value: array, size 2
+          0x2, 0x4,
+          // value: "en"
+          0x42, 0x65, 0x6e,
+          // value: "zh"
+          0x42, 0x7a, 0x68,
+        ],
+        name: 'map with array value',
+      },
+    ];
+
+    for (const tc of testCases) {
+      it(`should decode ${tc.name} correctly`, () => {
+        const decoder = new Decoder(Buffer.from(tc.input));
+        assert.deepStrictEqual(decoder.decode(0).value, tc.expected);
+      });
+    }
+  });
+
+  describe('decodeString()', () => {
+    const testCases = [
+      { expected: '', input: [0x40] },
+      { expected: '1', input: [0x41, 0x31] },
+      { expected: '人', input: [0x43, 0xe4, 0xba, 0xba] },
+      { expected: '123', input: [0x43, 0x31, 0x32, 0x33] },
+      {
+        expected: '123456789012345678901234567',
+        input: [
+          0x5b, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30,
+          0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30, 0x31,
+          0x32, 0x33, 0x34, 0x35, 0x36, 0x37,
+        ],
+      },
+      {
+        expected: '1234567890123456789012345678',
+        input: [
+          0x5c, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30,
+          0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30, 0x31,
+          0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38,
+        ],
+      },
+      {
+        expected: '12345678901234567890123456789',
+        input: [
+          0x5d, 0x00, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39,
+          0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30,
+          0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39,
+        ],
+      },
+      {
+        expected: '123456789012345678901234567890',
+        input: [
+          0x5d, 0x01, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39,
+          0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30,
+          0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39, 0x30,
+        ],
+      },
+      {
+        expected: 'x'.repeat(500),
+        input: [0x5e, 0x01, 0xc1, ...Array(500).fill(0x78)],
+      },
+      {
+        expected: 'x'.repeat(2000),
+        input: [0x5e, 0x06, 0xb4, ...Array(2000).fill(0x78)],
+      },
+      {
+        expected: 'x'.repeat(70000),
+        input: [0x5f, 0x01, 0x10, 0x30, ...Array(70000).fill(0x78)],
+      },
+    ];
+
+    for (const tc of testCases) {
+      const inputStr =
```

Comment:
This is used a bunch of times - maybe a worthwhile making a small helper fn for it. Same for `expectedStr`.